adserver/models.py

```diff
@@ -631,33 +639,42 @@ def clicks_needed_today(self):
 
         return self.clicks_remaining()
 
-    def weighted_clicks_needed_today(self):
+    def weighted_clicks_needed_today(self, publisher=None):
         """
         Calculates clicks needed taking into account a flight's priority.
 
         For the purpose of clicks needed, 1000 impressions = 1 click (for CPM ads)
         Takes into account value of the flight,
         which causes higher paid and better CTR ads to be prioritized.
+        Uses the passed publisher for a better CTR estimate if passed.
         """
         impressions_needed = 0
 
         # This is naive but we are counting a click as being worth 1,000 views
         impressions_needed += math.ceil(self.views_needed_today() / 1000.0)
         impressions_needed += self.clicks_needed_today()
 
-        # Cost should be a float from around .5-5, averaging 2
-        cost = float(self.cpm or self.cpc)
+        # Use the publisher CTR if available
+        # Otherwise, use this flight's average CTR
+        estimated_ctr = float(self.ctr())
```

Comment:
Seems this block should be in the `if self.cpc` block, since it's only used in that case, and makes the logic clearer.

I might even suggest an `if self.cpc` and `elif self.cpm` blocks to make it really obvious what we're doing.